modules/core/include/opencv2/core/hal/intrin_neon.hpp

```diff
@@ -78,8 +78,15 @@ CV_CPU_OPTIMIZATION_HAL_NAMESPACE_BEGIN
 #define CV_NEON_AARCH64 0
 #endif
 
-// TODO
+// Check if dot product intrinsics are supported as indicated by the
+// Arm C Language Extensions (ACLE) [1].
+//
+// [1] https://developer.arm.com/documentation/101028/0012/13--Advanced-SIMD--Neon--intrinsics
+#if __ARM_FEATURE_DOTPROD
+#define CV_NEON_DOT 1
```

Comment:
This line, defining macro `CV_NEON_DOT` is the ONLY thing that is done for M1.
Other things are NOT specific for M1, just not well optimized Armv8 NEON intrinsics.

So honestly, I prefer to separate this PR in 2 different patches.
Enabling M1 specific, following my previous comment, and the other huge modification for general Armv8.

Reply:
We've provided an updated PR with these changes. We will make a new PR based on this changeset that will make the CV_NEON_DOT changes for M1.